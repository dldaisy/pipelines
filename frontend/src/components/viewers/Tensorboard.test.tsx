/*
 * Copyright 2018 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from 'react';
import TensorboardViewer from './Tensorboard';
import TestUtils from '../../TestUtils';
import { Apis } from '../../lib/Apis';
import { PlotType } from './Viewer';
import { shallow, mount } from 'enzyme';

beforeEach(() => {
  jest.restoreAllMocks();
  jest.clearAllMocks();
});

describe('Tensorboard', () => {
  it('does not break on no config', () => {
    const tree = shallow(<TensorboardViewer configs={[]} />);
    expect(tree).toMatchSnapshot();
  });

  it('does not break on empty data', () => {
    const config = { type: PlotType.TENSORBOARD, url: '' };
    const tree = shallow(<TensorboardViewer configs={[config]} />);
    expect(tree).toMatchSnapshot();
  });

  it('does not break on empty data', () => {
    const config = { type: PlotType.TENSORBOARD, url: 'http://test/url' };
    const tree = shallow(<TensorboardViewer configs={[config]} />);
    expect(tree).toMatchSnapshot();
  });

  it('shows a link to the tensorboard instance if exists', async () => {
    const config = { type: PlotType.TENSORBOARD, url: 'http://test/url' };
    const mockGetApp = () => Promise.resolve('test/address');
    jest.spyOn(Apis, 'getTensorboardApp').mockImplementationOnce(mockGetApp);
    const tree = shallow(<TensorboardViewer configs={[config]} />);
    await mockGetApp;
    await TestUtils.flushPromises();
    expect(tree).toMatchSnapshot();
  });

  it('shows start button if no instance exists', async () => {
    const config = { type: PlotType.TENSORBOARD, url: 'http://test/url' };
    const defaultVersion = '1.14.0';
    const getAppMock = () => Promise.resolve('');
    const spy = jest.spyOn(Apis, 'getTensorboardApp').mockImplementationOnce(getAppMock);
    const tree = shallow(<TensorboardViewer configs={[config]} />);
    await getAppMock;
    await TestUtils.flushPromises();
    expect(tree).toMatchSnapshot();
    expect(spy).toHaveBeenCalledWith(config.url, defaultVersion);
  });

  it('starts tensorboard instance when button is clicked', async () => {
    const config = { type: PlotType.TENSORBOARD, url: 'http://test/url' };
    const getAppMock = () => Promise.resolve('');
    const defaultVersion = '1.14.0';
    const startAppMock = jest.fn(() => Promise.resolve(''));
    jest.spyOn(Apis, 'getTensorboardApp').mockImplementationOnce(getAppMock);
    jest.spyOn(Apis, 'startTensorboardApp').mockImplementationOnce(startAppMock);
    const tree = shallow(<TensorboardViewer configs={[config]} />);
    jest.spyOn(tree.instance(), 'setState').mockImplementation((_, cb) => cb && cb());
    await getAppMock;
    tree.find('BusyButton').simulate('click');
    await startAppMock;
    expect(startAppMock).toHaveBeenCalledWith('http%3A%2F%2Ftest%2Furl', defaultVersion);
  });

  it('starts tensorboard instance for two configs', async () => {
    const config = { type: PlotType.TENSORBOARD, url: 'http://test/url' };
    const config2 = { type: PlotType.TENSORBOARD, url: 'http://test/url2' };
    const defaultVersion = '1.14.0';
    const getAppMock = jest.fn(() => Promise.resolve(''));
    const startAppMock = jest.fn(() => Promise.resolve(''));
    jest.spyOn(Apis, 'getTensorboardApp').mockImplementationOnce(getAppMock);
    jest.spyOn(Apis, 'startTensorboardApp').mockImplementationOnce(startAppMock);
    const tree = shallow(<TensorboardViewer configs={[config, config2]} />);
    jest.spyOn(tree.instance(), 'setState').mockImplementation((_, cb) => cb && cb());
    await getAppMock;
    expect(getAppMock).toHaveBeenCalledWith(
      `Series1:${config.url},Series2:${config2.url}`,
      defaultVersion,
    );
    tree.find('BusyButton').simulate('click');
    await startAppMock;
    const expectedUrl =
      `Series1${encodeURIComponent(':' + config.url)}` +
      `${encodeURIComponent(',')}` +
      `Series2${encodeURIComponent(':' + config2.url)}`;
    expect(startAppMock).toHaveBeenCalledWith(expectedUrl, defaultVersion);
  });

  it('returns friendly display name', () => {
    expect(TensorboardViewer.prototype.getDisplayName()).toBe('Tensorboard');
  });

  it('is aggregatable', () => {
    expect(TensorboardViewer.prototype.isAggregatable()).toBeTruthy();
  });

  it('set state', async () => {
    const config = { type: PlotType.TENSORBOARD, url: 'http://test/url' };
    const tree = shallow(<TensorboardViewer configs={[config]} />);
    const event = { target: { value: '2.0.0' } } as React.ChangeEvent<{
      name?: string;
      value: unknown;
    }>;

    const onChangeMock = jest.fn(event => {
      tree.setState({ tensorflowVersion: event.target.value });
    });
    jest.spyOn(TensorboardViewer.prototype, 'onChangeFunc').mockImplementation(onChangeMock);
    TensorboardViewer.prototype.onChangeFunc(event);
    expect(onChangeMock).toHaveBeenCalled();
    expect(tree.state('tensorflowVersion')).toEqual('2.0.0');

    // then call the remote api with new state version
    const getAppMock = () => Promise.resolve('');
    const startAppMock = jest.fn(() => Promise.resolve(''));
    jest.spyOn(Apis, 'getTensorboardApp').mockImplementationOnce(getAppMock);
    jest.spyOn(Apis, 'startTensorboardApp').mockImplementationOnce(startAppMock);
    jest.spyOn(tree.instance(), 'setState').mockImplementation((_, cb) => cb && cb());
    await getAppMock;
    tree.find('BusyButton').simulate('click');
    await startAppMock;
    expect(startAppMock).toHaveBeenCalledWith(
      'http%3A%2F%2Ftest%2Furl',
      tree.state('tensorflowVersion'),
    );
  });

  it.only('simulate select', async () => {
    const getAppMock = () => Promise.resolve('');
    jest.spyOn(Apis, 'getTensorboardApp').mockImplementationOnce(getAppMock);
    const config = { type: PlotType.TENSORBOARD, url: 'http://test/url' };
    const spy = jest.spyOn(Apis, 'startTensorboardApp');

    const tree = mount(<TensorboardViewer configs={[config]} />);
    await TestUtils.flushPromises();

    tree
      .find('Select')
      .find('[role="button"]')
      .simulate('click');
    tree
      .findWhere(el => el.text() === 'TensorFlow 2.0.0')
      .hostNodes()
      .simulate('click');
    tree.find('BusyButton').simulate('click');
    expect(spy).toHaveBeenCalledWith('http%3A%2F%2Ftest%2Furl', '2.0.0');
  });
});
