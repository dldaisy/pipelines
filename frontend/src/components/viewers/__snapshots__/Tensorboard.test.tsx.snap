// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tensorboard does not break on empty data 1`] = `
<div>
  <div>
    <div
      className=""
    >
      <WithStyles(FormControl)
        className="Launch Tensorboard"
        style={
          Object {
            "minWidth": 120,
          }
        }
      >
        <WithStyles(WithFormControlContext(InputLabel))
          htmlFor="grouped-select"
        >
          TF Version
        </WithStyles(WithFormControlContext(InputLabel))>
        <WithStyles(WithFormControlContext(Select))
          input={
            <WithStyles(Input)
              id="grouped-select"
            />
          }
          onChange={[Function]}
          style={
            Object {
              "minHeight": 40,
            }
          }
          value="1.14.0"
        >
          <WithStyles(ListSubheader)>
            Tensoflow 1.x
          </WithStyles(ListSubheader)>
          <WithStyles(MenuItem)
            value="1.4.0"
          >
            TensorFlow 1.4.0
          </WithStyles(MenuItem)>
          <WithStyles(MenuItem)
            value="1.5.0"
          >
            TensorFlow 1.5.0
          </WithStyles(MenuItem)>
          <WithStyles(MenuItem)
            value="1.6.0"
          >
            TensorFlow 1.6.0
          </WithStyles(MenuItem)>
          <WithStyles(MenuItem)
            value="1.7.0"
          >
            TensorFlow 1.7.0
          </WithStyles(MenuItem)>
          <WithStyles(MenuItem)
            value="1.8.0"
          >
            TensorFlow 1.8.0
          </WithStyles(MenuItem)>
          <WithStyles(MenuItem)
            value="1.9.0"
          >
            TensorFlow 1.9.0
          </WithStyles(MenuItem)>
          <WithStyles(MenuItem)
            value="1.10.0"
          >
            TensorFlow 1.10.0
          </WithStyles(MenuItem)>
          <WithStyles(MenuItem)
            value="1.11.0"
          >
            TensorFlow 1.11.0
          </WithStyles(MenuItem)>
          <WithStyles(MenuItem)
            value="1.12.0"
          >
            TensorFlow 1.12.0
          </WithStyles(MenuItem)>
          <WithStyles(MenuItem)
            value="1.13.2"
          >
            TensorFlow 1.13.2
          </WithStyles(MenuItem)>
          <WithStyles(MenuItem)
            value="1.14.0"
          >
            TensorFlow 1.14.0
          </WithStyles(MenuItem)>
          <WithStyles(MenuItem)
            value="1.15.0"
          >
            TensorFlow 1.15.0
          </WithStyles(MenuItem)>
          <WithStyles(ListSubheader)>
            TensorFlow 2.x
          </WithStyles(ListSubheader)>
          <WithStyles(MenuItem)
            value="2.0.0"
          >
            TensorFlow 2.0.0
          </WithStyles(MenuItem)>
        </WithStyles(WithFormControlContext(Select))>
      </WithStyles(FormControl)>
    </div>
    <div>
      <BusyButton
        busy={true}
        className="buttonAction"
        onClick={[Function]}
        title="Start Tensorboard"
      />
    </div>
  </div>
</div>
`;

exports[`Tensorboard does not break on empty data 2`] = `
<div>
  <div>
    <div
      className=""
    >
      <WithStyles(FormControl)
        className="Launch Tensorboard"
        style={
          Object {
            "minWidth": 120,
          }
        }
      >
        <WithStyles(WithFormControlContext(InputLabel))
          htmlFor="grouped-select"
        >
          TF Version
        </WithStyles(WithFormControlContext(InputLabel))>
        <WithStyles(WithFormControlContext(Select))
          input={
            <WithStyles(Input)
              id="grouped-select"
            />
          }
          onChange={[Function]}
          style={
            Object {
              "minHeight": 40,
            }
          }
          value="1.14.0"
        >
          <WithStyles(ListSubheader)>
            Tensoflow 1.x
          </WithStyles(ListSubheader)>
          <WithStyles(MenuItem)
            value="1.4.0"
          >
            TensorFlow 1.4.0
          </WithStyles(MenuItem)>
          <WithStyles(MenuItem)
            value="1.5.0"
          >
            TensorFlow 1.5.0
          </WithStyles(MenuItem)>
          <WithStyles(MenuItem)
            value="1.6.0"
          >
            TensorFlow 1.6.0
          </WithStyles(MenuItem)>
          <WithStyles(MenuItem)
            value="1.7.0"
          >
            TensorFlow 1.7.0
          </WithStyles(MenuItem)>
          <WithStyles(MenuItem)
            value="1.8.0"
          >
            TensorFlow 1.8.0
          </WithStyles(MenuItem)>
          <WithStyles(MenuItem)
            value="1.9.0"
          >
            TensorFlow 1.9.0
          </WithStyles(MenuItem)>
          <WithStyles(MenuItem)
            value="1.10.0"
          >
            TensorFlow 1.10.0
          </WithStyles(MenuItem)>
          <WithStyles(MenuItem)
            value="1.11.0"
          >
            TensorFlow 1.11.0
          </WithStyles(MenuItem)>
          <WithStyles(MenuItem)
            value="1.12.0"
          >
            TensorFlow 1.12.0
          </WithStyles(MenuItem)>
          <WithStyles(MenuItem)
            value="1.13.2"
          >
            TensorFlow 1.13.2
          </WithStyles(MenuItem)>
          <WithStyles(MenuItem)
            value="1.14.0"
          >
            TensorFlow 1.14.0
          </WithStyles(MenuItem)>
          <WithStyles(MenuItem)
            value="1.15.0"
          >
            TensorFlow 1.15.0
          </WithStyles(MenuItem)>
          <WithStyles(ListSubheader)>
            TensorFlow 2.x
          </WithStyles(ListSubheader)>
          <WithStyles(MenuItem)
            value="2.0.0"
          >
            TensorFlow 2.0.0
          </WithStyles(MenuItem)>
        </WithStyles(WithFormControlContext(Select))>
      </WithStyles(FormControl)>
    </div>
    <div>
      <BusyButton
        busy={true}
        className="buttonAction"
        onClick={[Function]}
        title="Start Tensorboard"
      />
    </div>
  </div>
</div>
`;

exports[`Tensorboard does not break on no config 1`] = `
<div>
  <div>
    <div
      className=""
    >
      <WithStyles(FormControl)
        className="Launch Tensorboard"
        style={
          Object {
            "minWidth": 120,
          }
        }
      >
        <WithStyles(WithFormControlContext(InputLabel))
          htmlFor="grouped-select"
        >
          TF Version
        </WithStyles(WithFormControlContext(InputLabel))>
        <WithStyles(WithFormControlContext(Select))
          input={
            <WithStyles(Input)
              id="grouped-select"
            />
          }
          onChange={[Function]}
          style={
            Object {
              "minHeight": 40,
            }
          }
          value="1.14.0"
        >
          <WithStyles(ListSubheader)>
            Tensoflow 1.x
          </WithStyles(ListSubheader)>
          <WithStyles(MenuItem)
            value="1.4.0"
          >
            TensorFlow 1.4.0
          </WithStyles(MenuItem)>
          <WithStyles(MenuItem)
            value="1.5.0"
          >
            TensorFlow 1.5.0
          </WithStyles(MenuItem)>
          <WithStyles(MenuItem)
            value="1.6.0"
          >
            TensorFlow 1.6.0
          </WithStyles(MenuItem)>
          <WithStyles(MenuItem)
            value="1.7.0"
          >
            TensorFlow 1.7.0
          </WithStyles(MenuItem)>
          <WithStyles(MenuItem)
            value="1.8.0"
          >
            TensorFlow 1.8.0
          </WithStyles(MenuItem)>
          <WithStyles(MenuItem)
            value="1.9.0"
          >
            TensorFlow 1.9.0
          </WithStyles(MenuItem)>
          <WithStyles(MenuItem)
            value="1.10.0"
          >
            TensorFlow 1.10.0
          </WithStyles(MenuItem)>
          <WithStyles(MenuItem)
            value="1.11.0"
          >
            TensorFlow 1.11.0
          </WithStyles(MenuItem)>
          <WithStyles(MenuItem)
            value="1.12.0"
          >
            TensorFlow 1.12.0
          </WithStyles(MenuItem)>
          <WithStyles(MenuItem)
            value="1.13.2"
          >
            TensorFlow 1.13.2
          </WithStyles(MenuItem)>
          <WithStyles(MenuItem)
            value="1.14.0"
          >
            TensorFlow 1.14.0
          </WithStyles(MenuItem)>
          <WithStyles(MenuItem)
            value="1.15.0"
          >
            TensorFlow 1.15.0
          </WithStyles(MenuItem)>
          <WithStyles(ListSubheader)>
            TensorFlow 2.x
          </WithStyles(ListSubheader)>
          <WithStyles(MenuItem)
            value="2.0.0"
          >
            TensorFlow 2.0.0
          </WithStyles(MenuItem)>
        </WithStyles(WithFormControlContext(Select))>
      </WithStyles(FormControl)>
    </div>
    <div>
      <BusyButton
        busy={true}
        className="buttonAction"
        onClick={[Function]}
        title="Start Tensorboard"
      />
    </div>
  </div>
</div>
`;

exports[`Tensorboard shows a link to the tensorboard instance if exists 1`] = `
<div>
  <div>
    <div
      className=""
    >
      Tensorboard is running for this output.
    </div>
    <a
      className="unstyled"
      href="apis/v1beta1/_proxy/test/address"
      target="_blank"
    >
      <WithStyles(Button)
        className="buttonAction"
        disabled={false}
        style={
          Object {
            "marginBottom": 20,
            "width": 150,
          }
        }
      >
        Open Tensorboard
      </WithStyles(Button)>
    </a>
    <div>
      <WithStyles(Button)
        className="buttonAction"
        disabled={false}
        id="delete"
        onClick={[Function]}
        style={
          Object {
            "marginBottom": 20,
            "width": 150,
          }
        }
        title="stop tensorboard and delete its instance"
      >
        Delete Tensorboard
      </WithStyles(Button)>
      <WithStyles(Dialog)
        aria-labelledby="dialog-title"
        onClose={[Function]}
        open={false}
      >
        <WithStyles(DialogTitle)
          id="dialog-title"
          style={
            Object {
              "cursor": "move",
            }
          }
        >
          Stop Tensorboard 1.14.0?
        </WithStyles(DialogTitle)>
        <WithStyles(DialogContent)>
          <WithStyles(DialogContentText)>
            You can stop the current running tensorboard. The tensorboard viewer will also be clean from your GKE workloads.
          </WithStyles(DialogContentText)>
        </WithStyles(DialogContent)>
        <WithStyles(DialogActions)>
          <WithStyles(Button)
            autoFocus={true}
            color="primary"
            onClick={[Function]}
            style={
              Object {
                "width": 50,
              }
            }
          >
            Cancel
          </WithStyles(Button)>
          <BusyButton
            busy={false}
            className="buttonAction"
            color="primary"
            onClick={[Function]}
            style={
              Object {
                "width": 50,
              }
            }
            title="Stop"
          />
        </WithStyles(DialogActions)>
      </WithStyles(Dialog)>
    </div>
  </div>
</div>
`;

exports[`Tensorboard shows start button if no instance exists 1`] = `
<div>
  <div>
    <div
      className=""
    >
      <WithStyles(FormControl)
        className="Launch Tensorboard"
        style={
          Object {
            "minWidth": 120,
          }
        }
      >
        <WithStyles(WithFormControlContext(InputLabel))
          htmlFor="grouped-select"
        >
          TF Version
        </WithStyles(WithFormControlContext(InputLabel))>
        <WithStyles(WithFormControlContext(Select))
          input={
            <WithStyles(Input)
              id="grouped-select"
            />
          }
          onChange={[Function]}
          style={
            Object {
              "minHeight": 40,
            }
          }
          value="1.14.0"
        >
          <WithStyles(ListSubheader)>
            Tensoflow 1.x
          </WithStyles(ListSubheader)>
          <WithStyles(MenuItem)
            value="1.4.0"
          >
            TensorFlow 1.4.0
          </WithStyles(MenuItem)>
          <WithStyles(MenuItem)
            value="1.5.0"
          >
            TensorFlow 1.5.0
          </WithStyles(MenuItem)>
          <WithStyles(MenuItem)
            value="1.6.0"
          >
            TensorFlow 1.6.0
          </WithStyles(MenuItem)>
          <WithStyles(MenuItem)
            value="1.7.0"
          >
            TensorFlow 1.7.0
          </WithStyles(MenuItem)>
          <WithStyles(MenuItem)
            value="1.8.0"
          >
            TensorFlow 1.8.0
          </WithStyles(MenuItem)>
          <WithStyles(MenuItem)
            value="1.9.0"
          >
            TensorFlow 1.9.0
          </WithStyles(MenuItem)>
          <WithStyles(MenuItem)
            value="1.10.0"
          >
            TensorFlow 1.10.0
          </WithStyles(MenuItem)>
          <WithStyles(MenuItem)
            value="1.11.0"
          >
            TensorFlow 1.11.0
          </WithStyles(MenuItem)>
          <WithStyles(MenuItem)
            value="1.12.0"
          >
            TensorFlow 1.12.0
          </WithStyles(MenuItem)>
          <WithStyles(MenuItem)
            value="1.13.2"
          >
            TensorFlow 1.13.2
          </WithStyles(MenuItem)>
          <WithStyles(MenuItem)
            value="1.14.0"
          >
            TensorFlow 1.14.0
          </WithStyles(MenuItem)>
          <WithStyles(MenuItem)
            value="1.15.0"
          >
            TensorFlow 1.15.0
          </WithStyles(MenuItem)>
          <WithStyles(ListSubheader)>
            TensorFlow 2.x
          </WithStyles(ListSubheader)>
          <WithStyles(MenuItem)
            value="2.0.0"
          >
            TensorFlow 2.0.0
          </WithStyles(MenuItem)>
        </WithStyles(WithFormControlContext(Select))>
      </WithStyles(FormControl)>
    </div>
    <div>
      <BusyButton
        busy={false}
        className="buttonAction"
        onClick={[Function]}
        title="Start Tensorboard"
      />
    </div>
  </div>
</div>
`;
